toss/fitness/fitness_function_utils.py

```diff
@@ -45,4 +45,66 @@ def _compute_squared_distance(positions,constant):
     Returns:
         (np.ndarray): (N,) array containing average distance from each point in arr to constant.
     """
-    return np.sum(np.power(positions,2), axis=0) - constant**2
\ No newline at end of file
+    return np.sum(np.power(positions,2), axis=0) - constant**2
+
+
+def compute_space_coverage(positions, radius_outer_bounding_sphere):
+    """
+    In this function, we create a spherical meshgrid by defining a number of points
+    inside the outer bounding sphere. We then generate a multidimensional array
+    of boolean values corresponding to each point in the meshgrid. These values 
+    are initialy set to false since they have not been visited by any spacecraft.
+
+    Given a set of posiions on the trajectory that are defined inside the 
+    outer bounding sphere, we identify the points that are the closest the given positions, 
+    and subsequently reassign the visited boolean value to true. The ratio of visited points 
+    is then given by the number of True values to the total number of values in the boolean array.
+
+    Args:
+        positions (_type_): (3,N) Array of positions (in cartesian coordinates).
+        radius_outer_bounding_sphere (float): Radius of outer bounding sphere.
+
+    Returns:
+        ratio (float): Number of True values to the total number of values in the boolean array
+    """
+    # Set the detail of the spherical meshgrid:
+    r_num = 1000 # Number of evenly spaced points along the radial axis
+    theta_num = 1450 # Number of evenly spaced points along the polar angle (defined in between 0 and pi)
+    phi_num = 1450 # Number of evenly spaced points along the azimuthal angle (defined in between 0 and 2*pi)
+
+    # Create a spherical meshgrid
+    r = np.linspace(0, radius_outer_bounding_sphere, r_num) # radial coordinates
+    theta = np.linspace(0, np.pi, theta_num) # polar angle coordinates
+    phi = np.linspace(0, 2*np.pi, phi_num) # azimuthal angle coordinates
+    r_matrix, theta_matrix, phi_matrix = np.meshgrid(r, theta, phi) # 3D arrays of r, theta and phi
+    
+    # Create a boolean array with the same shape as the spherical meshgrid
+    bool_array = np.zeros_like(r_matrix, dtype=bool) # initialize with False
+
+    # Create grid of positions
+    x = positions[0,:]
+    y = positions[1,:]
+    z = positions[2,:]
+    x2, y2, z2 = np.meshgrid(x, y, z, indexing='ij')
+    all_grid = np.array([x2.flatten(), y2.flatten(), z2.flatten()]).T
```

Comment:
Why do you create a grid from the positions? This makes no sense. You should be just iterating over the points of the trajectory and checking in which of the grid cells they are and then flipping those.